File: macos/Sources/dfdaemon/Stats.swift
import Foundation

/// Query-count stats only — never the actual queried domains. Matches the
/// project's "category matches, not browsing history" privacy stance.
final class Stats {
    private let lock = NSLock()
    private var totalQueries = 0
    private var blockedAlwaysOn = 0
    private var blockedSocial = 0
    private var safeSearchRewrites = 0
    private var lastWrite = Date.distantPast
    private let writeInterval: TimeInterval = 10

    func recordTotal() {
        lock.lock(); totalQueries += 1; lock.unlock()
        writeIfDue()
    }

    func recordBlockedAlwaysOn() {
        lock.lock(); blockedAlwaysOn += 1; lock.unlock()
    }

    func recordBlockedSocial() {
        lock.lock(); blockedSocial += 1; lock.unlock()
    }

    func recordSafeSearchRewrite() {
        lock.lock(); safeSearchRewrites += 1; lock.unlock()
    }

    private func writeIfDue() {
        let now = Date()
        lock.lock()
        guard now.timeIntervalSince(lastWrite) >= writeInterval else { lock.unlock(); return }
        lastWrite = now
        let snapshot: [String: Any] = [
            "totalQueries": totalQueries,
            "blockedAlwaysOn": blockedAlwaysOn,
            "blockedSocial": blockedSocial,
            "safeSearchRewrites": safeSearchRewrites,
            "updatedAt": ISO8601DateFormatter().string(from: now),
        ]
        lock.unlock()
        guard let data = try? JSONSerialization.data(withJSONObject: snapshot, options: [.prettyPrinted]) else { return }
        try? data.write(to: URL(fileURLWithPath: Config.statsFile), options: .atomic)
    }
}

File: android/app/src/main/java/com/distractionfree/app/Stats.kt
package com.distractionfree.app

import android.content.Context
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicInteger

/** Query-count stats only — never the actual queried domains. */
class Stats(private val context: Context) {
    private val total = AtomicInteger(0)
    private val blockedAlwaysOn = AtomicInteger(0)
    private val blockedSocial = AtomicInteger(0)
    private val safeSearchRewrites = AtomicInteger(0)
    @Volatile private var lastWrite = 0L
    private val writeIntervalMs = 10_000L

    fun recordTotal() { total.incrementAndGet(); writeIfDue() }
    fun recordBlockedAlwaysOn() = blockedAlwaysOn.incrementAndGet()
    fun recordBlockedSocial() = blockedSocial.incrementAndGet()
    fun recordSafeSearchRewrite() = safeSearchRewrites.incrementAndGet()

    private fun writeIfDue() {
        val now = System.currentTimeMillis()
        if (now - lastWrite < writeIntervalMs) return
        lastWrite = now
        val obj = JSONObject()
        obj.put("totalQueries", total.get())
        obj.put("blockedAlwaysOn", blockedAlwaysOn.get())
        obj.put("blockedSocial", blockedSocial.get())
        obj.put("safeSearchRewrites", safeSearchRewrites.get())
        obj.put("updatedAt", now)
        try {
            AppPaths.statsFile(context).writeText(obj.toString())
        } catch (e: Exception) {
            // best-effort
        }
    }
}
